lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const catchUpResponseTimeout = 5 * time.Second
+
+type catchUp struct {
+	lock sync.Mutex // applied on requestsSent
+
+	requestsSent map[peer.ID]CatchUpRequest
+	bestResponse *atomic.Value // *CatchUpResponse
+
+	catchUpResponseCh chan *CatchUpResponse
+	waitingOnResponse *atomic.Value
+	grandpa           *Service
+}
+
+func (c *catchUp) do(to peer.ID, round uint64, setID uint64) error {
+	defer c.waitingOnResponse.Store(false)
+
+	if err := c.sendCatchUpRequest(
+		to, newCatchUpRequest(round, setID),
+	); err != nil {
+		return fmt.Errorf("failed to send catch up request: %w", err)
+	}
+
+	logger.Debugf("successfully sent a catch up request to node %s, for round number %d and set ID %d",
+		to, round, setID)
+
+	c.grandpa.paused.Store(true)
+
+	timer := time.NewTimer(catchUpResponseTimeout)
+	defer timer.Stop()
+
+	select {
+	case <-c.catchUpResponseCh:
+		return nil
+	case <-timer.C:
+		return errors.New("timeout")
+	}
+}
+
+func (c *catchUp) sendCatchUpRequest(to peer.ID, req *CatchUpRequest) error {
+	if c.bestResponse.Load() != nil {
+		logger.Debug("ignoring neighbour message since we are already processing a catch-up response")
+		return nil
+	}
+
+	// TODO: Clean up all request sent before 5 min / (neighbour message interval)
+	c.lock.Lock()
+	_, ok := c.requestsSent[to]
+	c.lock.Unlock()
+	if ok {
+		logger.Debugf("ignoring neighbour message since we already sent a catch-up request to this peer: %s", to)
+		return nil
+	}
+
+	c.waitingOnResponse.Store(true)
+
+	cm, err := req.ToConsensusMessage()
+	if err != nil {
+		return fmt.Errorf("cannot convert request to consensus message: %w", err)
+	}
+
+	err = c.grandpa.network.SendMessage(to, cm)
+	if err != nil {
+		return fmt.Errorf("cannot send grandpa message: %w", err)
+	}
+
+	c.lock.Lock()
+	c.requestsSent[to] = *req
+	c.lock.Unlock()
+	c.grandpa.paused.Store(true)
+
+	return nil
+}
+
+func (c *catchUp) handleCatchUpResponse(msg *CatchUpResponse) error {
+	if !c.grandpa.authority {
+		return nil
+	}
+
+	logger.Debugf(
+		"processing catch up response with hash %s for round %d and set id %d",
+		msg.Hash, msg.Round, msg.SetID)
+
+	// if we aren't currently expecting a catch up response, return
+	if !c.grandpa.paused.Load().(bool) {
```

Comment:
Should `c.waitingOnResponse` be used instead?

Reply:
replaced this with `c.waitingOnResponse`. I don't think `grandpa.paused` is useful anymore. Removed it.